### PR TITLE
Cap RT-DETR denoising queries at the query budget

### PR DESCRIPTION
**Pull request by artest08**
An image with 700 boxes emitted 1400 denoising queries against a budget of 200, growing decoder self-attention quadratically.

On datasets whose images carry more boxes than `num_dn`, `get_cdn_group` sized the denoising group from the largest image in the batch instead of from the budget, so the denoising block and the decoder self-attention mask grew with the box count, wasting memory and running training out of it on dense data. That cost is quadratic in the block size and paid again in every decoder layer, so capping the group at `num_dn` keeps the mask at the budgeted size and lets dense datasets train within memory and faster. Images above the budget denoise a uniformly random subset of their boxes redrawn on every call, so every box is still denoised over training, while images under the budget keep all of theirs.

```python
from ultralytics import RTDETR

RTDETR("rtdetr-l.yaml").train(data="SKU-110K.yaml", epochs=1, imgsz=640)
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
RT-DETR denoising queries are now capped at the configured `num_dn` budget by sampling oversized ground-truth sets before building the denoising block.

### 📊 Key Changes
- `get_cdn_group` limits the per-image denoising box count to `num_dn` and selects a new random subset when an image contains more boxes.
- Denoising groups and attention-mask dimensions are calculated from the capped count instead of the largest uncapped image.
- Denoising metadata now records the selected ground-truth indices through `dn_gt_idx`.
- `get_dn_match_indices` uses those selected indices to match denoising predictions with their corresponding targets.

### 🎯 Purpose & Impact
- Images with more than `num_dn` boxes no longer expand the denoising block and decoder self-attention mask beyond the configured budget; images within the budget retain all their boxes.
- The selected subset is regenerated on each call, while denoising loss matching remains aligned with the sampled targets.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @artest08, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

**Review by UltralyticsAssistant — APPROVED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

LGTM. Reviewed the denoising-query cap, random per-image subsampling, metadata propagation, and denoising loss matching. The sampled global ground-truth indices remain aligned with the unchanged batch targets, and the query/mask dimensions stay within the configured denoising budget. No concrete regressions found.

**Comment by UltralyticsAssistant** [comment]
🎉 PR [#25834](https://github.com/ultralytics/ultralytics/pull/25834) has been merged—thank you @artest08!

> “Great things are done by a series of small things brought together.” — Vincent van Gogh

This focused improvement keeps RT-DETR denoising queries within the configured `num_dn` budget while preserving aligned matching through the sampled ground-truth indices.
